### PR TITLE
First version memory pair game

### PR DESCRIPTION
**Pull request by Levi123**
### Memory - Pair Game

[Demo](https://levi123.github.io/Memory-Pair-Game/) |
[Code base](https://github.com/Levi123/Memory-Pair-Game)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

File: submissions/Levi123/memory-pair-game/script.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            id:6,
+            images: 'img/koala.png'
+        },
+    ]
+
+
+    function shuffle(arrayToShuffle) {
+        for(var j, x, i = arrayToShuffle.length; i; j = parseInt(Math.random() * i), x = arrayToShuffle[--i], arrayToShuffle[i] = arrayToShuffle[j], arrayToShuffle[j] = x);
+        return arrayToShuffle;
+    };
```
Comment (madmaxWMFU) [inline submissions/Levi123/memory-pair-game/script.js:58]: Read [shuffle post](https://dev.to/codebubb/how-to-shuffle-an-array-in-javascript-2ikj)

File: submissions/Levi123/memory-pair-game/script.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        let list = `
+            <div class="main-content__element flipper">
+                <div class="front">
+                    <img src="img/card.png" alt="" class="flip-card-img" data-id=${i.id}>
+                </div>
+                <div class="back">
+                    <img src=${i.images} alt="" class="flip-card-back-img">
+                </div>
+            </div>
+        `
```
Comment (madmaxWMFU) [inline submissions/Levi123/memory-pair-game/script.js:73]: Try methods `reduce` or `map`

File: submissions/Levi123/memory-pair-game/script.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    ]
+
+
+    function shuffle(arrayToShuffle) {
+        for(var j, x, i = arrayToShuffle.length; i; j = parseInt(Math.random() * i), x = arrayToShuffle[--i], arrayToShuffle[i] = arrayToShuffle[j], arrayToShuffle[j] = x);
+        return arrayToShuffle;
+    };
+
+    const generateElementForBoard = shuffle([...allElementsForGame, ...allElementsForGame]);
+    console.log(generateElementForBoard);
```
Comment (madmaxWMFU) [inline submissions/Levi123/memory-pair-game/script.js:61]: Delete a debug line.

File: submissions/Levi123/memory-pair-game/script.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+}       
+
+generateGame();
+
+const startGame = () => {
+    STATE.statusGame = true;
+    STATE.loop = setInterval(() => {
+        STATE.totalTime++
+        SELECTORS.timer.innerText = ` ${STATE.totalTime} sec`
+    }, 1000)
```
Comment (madmaxWMFU) [inline submissions/Levi123/memory-pair-game/script.js:85]: A magic number.

File: submissions/Levi123/memory-pair-game/script.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    timer : document.querySelector('.seconds'),
+    flipCounter: document.querySelector('.flips'),
+}
+
+const STATE = {
+    totalFlips : 0,
+    totalTime : 0,
+    flipElementIdArray : [],
+    cardOne : undefined,
+    cardTwo : undefined,
```
Comment (madmaxWMFU) [inline submissions/Levi123/memory-pair-game/script.js:14]:
```suggestion
    cardOne : null,
    cardTwo : null,
```

**Review by madmaxWMFU — CHANGES_REQUESTED** [review]
Good work! See comments below.
Check your code in [validator](https://validator.w3.org/nu/?doc=https%3A%2F%2Flevi123.github.io%2FMemory-Pair-Game%2F)

File: submissions/Levi123/memory-pair-game/script.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
+const SELECTORS = {
+    flipContainer : document.querySelector('.flip-container'),
+    content : document.querySelector('.main-content__playing-field'),
+    textContent: document.querySelector('.main-content'),
+    timer : document.querySelector('.seconds'),
+    flipCounter: document.querySelector('.flips'),
+}
```
[the remark this answers is not in the document]
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js:7]: @madmaxWMFU done

File: submissions/Levi123/memory-pair-game/script.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    totalFlips : 0,
+    totalTime : 0,
+    flipElementIdArray : [],
+    cardOne : undefined,
+    cardTwo : undefined,
+    statusGame : false,
+    loop: null,
+    combinationCounter: 6,
+    isInteractive: true,
+}
```
[the remark this answers is not in the document]
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js:19]: @madmaxWMFU thanks!

File: submissions/Levi123/memory-pair-game/script.js (hunk 1 above)
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js:58]: @madmaxWMFU done

File: submissions/Levi123/memory-pair-game/script.js (hunk 2 above)
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js:73]: @madmaxWMFU used the forEach method, will it be appropriate?

File: submissions/Levi123/memory-pair-game/script.js (hunk 3 above)
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js:61]: @madmaxWMFU Done

File: submissions/Levi123/memory-pair-game/script.js (hunk 4 above)
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js:85]: @madmaxWMFU Don't understand

File: submissions/Levi123/memory-pair-game/script.js (hunk 5 above)
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js:14]: @madmaxWMFU thank you, I understood null it's - the value of the absence of an object, while undefined - uncertainty.

File: submissions/Levi123/memory-pair-game/script.js (hunk 4 above)
Comment (madmaxWMFU) [reply submissions/Levi123/memory-pair-game/script.js:85]: 1000 as variable

Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js]: > 1000 as variable

done

File: submissions/Levi123/memory-pair-game/script.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            </div>
+            <div class="back">
+                <img src=${elementForRender.images} alt="" class="flip-card-back-img">
+            </div>
+        </div>
+    `
+    })
+}       
+
+generateGame();
```
Comment (madmaxWMFU) [inline submissions/Levi123/memory-pair-game/script.js:79]: `window.addEventListener('DOMContentLoaded', generateGame);`

**Review by madmaxWMFU — CHANGES_REQUESTED** [review]

File: submissions/Levi123/memory-pair-game/script.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    }
+}
+ 
+flipContainer.addEventListener('click', function (event) {
+    const eventTarget = event.target;
+    const eventParent = eventTarget.parentElement;
+    if (event.target.closest('.front') && STATE.isInteractive){
+        flipCard(eventParent.parentElement);
+        checkCard(eventTarget, eventParent.parentElement);
+        if (STATE.totalFlips === 1){
```
[the remark this answers is not in the document]
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js]: > Also, a magic number

done

File: submissions/Levi123/memory-pair-game/script.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        STATE.isInteractive = false;
+        setTimeout(() => {
+            STATE.cardOne.classList.remove('rotate_card');
+            STATE.cardTwo.classList.remove('rotate_card');
+            STATE.cardOne = null;
+            STATE.isInteractive = true;
+        }, STATE.timeToAnimation)
+    }
+
+    if (STATE.combinationCounter === 0){
```
[the remark this answers is not in the document]
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js]: > 0 👀

done

File: submissions/Levi123/memory-pair-game/script.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        console.log(STATE.flipElementIdArray)
+    }
+    STATE.isInteractive = true;
+
+    if (!STATE.cardOne){
+        return STATE.cardOne = parentElement;
+    }
+    STATE.cardTwo = parentElement;
+
+    if (STATE.flipElementIdArray.length === 2 && STATE.flipElementIdArray[0] === STATE.flipElementIdArray[1]){
```
[the remark this answers is not in the document]
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js]: done

File: submissions/Levi123/memory-pair-game/script.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    STATE.totalFlips++;
+    target.classList.add('rotate_card');
+    flipCounter.innerText = ` ${STATE.totalFlips}`;
+}
+
+const checkCard = (target, parentElement) => {
+    if (STATE.flipElementIdArray.length < 2){
+        STATE.isInteractive = false;
+        STATE.flipElementIdArray.push(target.dataset.id);
+        console.log(STATE.flipElementIdArray)
```
[the remark this answers is not in the document]
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js]: > Delete debug line

done

File: submissions/Levi123/memory-pair-game/script.js (hunk 8 above)
Comment (Levi123) [reply submissions/Levi123/memory-pair-game/script.js]: > `window.addEventListener('DOMContentLoaded', generateGame);`

done

**Review by madmaxWMFU — APPROVED** [review]
Great, I approve of your work. Good luck with future tasks!

**Pull request merged**